### ISSUE TITLE
[Bug][Devlake-backend]  _raw_gitlab_api_job size too big

### ISSUE THREAD

**Issue by antoniomuso** [apache/devlake#6763]
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-devlake/issues?q=is%3Aissue) and found no similar issues.

### What happened

The size of the table  _raw_gitlab_api_job is too big. The disk size of this table is 75% of the total size given to the MySQL instance.

### What do you expect to happen

The size should not be too big.

### How to reproduce

Normal devlake use.

### Anything else

I wonder if this is a bug or if it is normal. I would like to know how to handle this if it is normal. Is there some way to clean up _raw tables after usage?

### Version

v0.18.0

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

**Comment by Startrekzky** [comment]
Hi @antoniomuso . May I ask
1. the size of your MySQL?
2. the actual number of GitLab job in your GitLab UI? You can check it on the Pipelines page. Get the iid number in the console and multiply it by the rough average number of jobs per pipeline to infer the job number. It might be a lot more than you can imagine.

![image](https://github.com/apache/incubator-devlake/assets/14050754/b644bbf7-78df-4d7e-8c25-ab13e9e1f109)

![image](https://github.com/apache/incubator-devlake/assets/14050754/2fddab81-341b-49fc-a0e6-27b505a8fd56)

With the above info, I will have a more educated judgement. Thanks.

**Comment by antoniomuso** [comment]
There is a discrepancy:
```sql
SELECT count(id)
FROM  lake._raw_gitlab_api_job
```
Return to me 4534738 entries. If I try to query the domain layer
```sql
WITH _t AS 
 (SELECT repo_id, count(t.id) as jobs 
 FROM lake.cicd_tasks t, lake.cicd_pipelines p, cicd_pipeline_commits c 
 WHERE t.pipeline_id = p.id AND c.pipeline_id = p.id 
 GROUP BY repo_id
 )

select sum(jobs) from _t 
```
 I obtain 240414 entries.

**Comment by antoniomuso** [comment]
Currently, the size of my MySQL is 18G.
The iid multiplied by the average of jobs per pipeline gives me 311101

**Comment by simonecorsi** [comment]
Hey,  I'm currently working to gain a deeper understanding of the system requirements, sizing considerations, and scalability aspects to run DevLake on our org GitLab instance, particularly in the context of managing a substantial amount of data and logs associated with projects spanning several years.

I've been examining the `_raw_gitlab_api_job` and noticed a discrepancy in the job count when comparing it with the parsed data. 

My understanding of the data model is that the primary key `id` mirrors the GitLab unique job id, so there should be 1:1 count, but as pointed by @antoniomuso there seems to be more for some reason. 

It's difficult for me, even picking random samples to find it on dataset this big, so I wanted to ask someone with a wider knowledge of DevLake internals.

I get the necessity to preserve raw data to avoid reprocessing (which is good), but I'm concerned about scalability, if this cannot be removed or cleaned and needs to be kept, we could consider this a long-term JSON bucket, would it then be viable to consider implementing data compression techniques for this JSON bucket? Several compression algorithms are available that could significantly optimize storage usage.

Thanks in advance for any clarification 🙏

**Comment by antoniomuso** [comment]
Hi, updates about this. `_raw_gitlab_api_job` continue too increase his size in my instance,  2 Weeks ago has had 4.534.738 entires and now it have 6.283.978 entries. We don't have all this jobs in the imported projects.

**Comment by klesh** [comment]
Hi, Sorry for replying so late.

# TLDR: upgrade to v0.19 or later


# Long version:
All tables start with `_raw` are designed for the Incremental Collection feature. It contains all historical records of an Entity from the API responses, in this case, the gitlab `job`. Historical means that for a specific `job`, we might collect it multiple times and the Latest collected record (with the highest `id` in the `_raw_github_api_jobs`) would end up in the `_tool_github_jobs` table.
For APIs that support filtering records by `updated at` timestamp, it works fine and won't produce too many duplicate records if any.
However, the Gitlab jobs API doesn't support such a filter which leads to the problem you all are facing.
So, I took another route to avoid such a problem https://github.com/apache/incubator-devlake/pull/5889

**Comment by klesh** [comment]
Remember to run a Full Sync after the upgrade to flush out the _raw data

**Comment by antoniomuso** [comment]
Thanks. I tried to resync using the `Collect all` button, but nothing happened. Then I manually called `DROP TABLE _raw_github_api_jobs`., And now seems to work.

**Comment by antoniomuso** [comment]
@klesh v0.19

**Comment by antoniomuso** [comment]
Also, I tried to clean the data scope, and nothing happened to _the raw table. @klesh If I drop the table every month, could there be some inconsistency in the data?

**Comment by klesh** [comment]
@antoniomuso Don't do that, there is different sync in play, and dropping _raw tables would cause data being incomplete.
@abeizn When did we introduce full sync?  does v0.19 support that?

**Comment by abeizn** [comment]
@klesh @antoniomuso v0.20 support full sync mode, you can upgrade it to [v0.20.0-beta10 ](https://github.com/apache/incubator-devlake/releases/tag/v0.20.0-beta10)

**Comment by simonecorsi** [comment]
> @antoniomuso Don't do that, there is different sync in play, and dropping _raw tables would cause data being incomplete. 

Let's say we have a data retention of 1year, if I want to clean up data older than one year each day to save resources will this lead to some kind of inconsistency? If so, what's an advisable way to handle the case?

**Comment by klesh** [comment]
@simonecorsi Data retention is not supported at this point. There was an issue discussing the matter but never gets enough traction. https://github.com/apache/incubator-devlake/issues/4621
